lib/Serialization/Deserialization.cpp

```diff
@@ -4257,6 +4283,36 @@ llvm::Error DeclDeserializer::deserializeDeclAttributes() {
         break;
       }
 
+      case decls_block::Differentiable_DECL_ATTR: {
+        bool isImplicit;
+        bool linear;
+        GenericSignatureID derivativeGenSigId;
+        ArrayRef<uint64_t> parameters;
+
+        serialization::decls_block::DifferentiableDeclAttrLayout::readRecord(
+            scratch, isImplicit, linear, derivativeGenSigId, parameters);
+
+        auto derivativeGenSig = MF.getGenericSignature(derivativeGenSigId);
+        llvm::SmallBitVector parametersBitVector(parameters.size());
+        for (unsigned i : indices(parameters))
+          parametersBitVector[i] = parameters[i];
+        auto *indices = IndexSubset::get(ctx, parametersBitVector);
+        auto *diffAttr = DifferentiableAttr::create(
+            ctx, isImplicit, SourceLoc(), SourceRange(), linear,
+            /*parsedParameters*/ {}, /*trailingWhereClause*/ nullptr);
+
+        // Cache parameter indices so that they can set later.
+        // `DifferentiableAttr::setParameterIndices` cannot be called here
+        // because it requires `DifferentiableAttr::setOriginalDeclaration` to
+        // be called first. `DifferentiableAttr::setOriginalDeclaration` cannot
+        // be called here because the original declaration is not accessible in
+        // this function (`DeclDeserializer::deserializeDeclAttributes`).
```

Comment:
Note: `DynamicReplacementAttr` has the same issue because it also stores the "original" (replaced function) declaration. However, it stores the `DeclID` of the original declaration instead of an actual `Decl *`: [definition](https://github.com/apple/swift/blob/1be86adbfc8f9c178e3bbfdc66340f7aa526a40b/include/swift/AST/Attr.h#L1042-L1043), [deserialization](https://github.com/apple/swift/blob/b8cd6cadeb24c635153d9fcf00e5adce518b13d4/lib/Serialization/Deserialization.cpp#L4217-L4233).

We could try adopting that approach to remove this deserialization cruft (`setOriginalDeclarationAndParameterIndicesInDifferentiableAttributes`). The approach sounds like a nice incremental improvement if it works.